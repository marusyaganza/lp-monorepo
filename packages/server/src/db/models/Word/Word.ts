import {
  SpacedRepetitionData,
  SpacedRepetitionMap,
  Word
} from '../../schema/Word';
import {
  calculatePaginationValues,
  GAME_FILTERS,
  getGameFilters,
  getGameSortingFilter,
  getTagsFilter,
  getTenseFilter,
  getWordSortingFilters,
  OPTIONS_PROJECTIONS,
  PROJECTIONS
} from '../../helpers';

import {
  Word as WordType,
  NewWordInput,
  UpdateStatisticsInput,
  Game as GameType,
  WordStatisticsField,
  PaginatedWords,
  UpdateWordInput,
  GameDataInput,
  SortBy,
  Language,
  WordsPerPageInput,
  SortWordsBy,
  Verb,
  GameSettings,
  Score,
  Game,
  Tense
} from '../../../generated/graphql';
import { getWordsFilters, filterTags } from '../../helpers';
import { UserInputError } from '../../../utils/apolloCustomErrors';
import { DEFAULT_OPTIONS__SAMPLE_SIZE } from '../../../constants/defaultValues';
import {
  initializeSpacedRepetitionFields,
  updateSpacedRepetitionFields
} from '../../helpers/general/spacedRepetition';
import { isConjugationSpacedRepetitionData } from '../../../types/typeGuards';

const STATISTICS_FIELD: WordStatisticsField = {
  lastTimePracticed: 0,
  practicedTimes: 0,
  errorCount: 0,
  successRate: 0
};

const DEFAULT_STATISTICS: Record<GameType, WordStatisticsField> = {
  [GameType.Audio]: STATISTICS_FIELD,
  [GameType.SelectDef]: STATISTICS_FIELD,
  [GameType.SelectWord]: STATISTICS_FIELD,
  [GameType.TypeWord]: STATISTICS_FIELD,
  [GameType.Conjugation]: STATISTICS_FIELD,
  [GameType.Image]: STATISTICS_FIELD,
  [GameType.Gender]: STATISTICS_FIELD,
  [GameType.Speaking]: STATISTICS_FIELD
};

const repetitionData = initializeSpacedRepetitionFields();

const CONJUGATION_REPETITION = {
  [Tense.Cond]: repetitionData,
  [Tense.Futr]: repetitionData,
  [Tense.Impf]: repetitionData,
  [Tense.Pind]: repetitionData,
  [Tense.Pprf]: repetitionData,
  [Tense.Pret]: repetitionData,
  [Tense.Psub]: repetitionData,
  [Tense.Ppci]: repetitionData
};

const DEFAULT_REPETITION_DATA: SpacedRepetitionMap = {
  [GameType.Audio]: repetitionData,
  [GameType.SelectDef]: repetitionData,
  [GameType.SelectWord]: repetitionData,
  [GameType.TypeWord]: repetitionData,
  [GameType.Image]: repetitionData,
  [GameType.Conjugation]: CONJUGATION_REPETITION,
  [GameType.Gender]: repetitionData,
  [GameType.Speaking]: repetitionData
};

export interface WordModelType {
  findById: (id: string, user: string) => Promise<WordType | null>;
  findOne: (
    filter: Partial<WordType>,
    user: string
  ) => Promise<WordType | null | undefined>;
  findManyAndPaginate: (
    filter: WordsPerPageInput | null | undefined,
    user: string
  ) => Promise<PaginatedWords>;
  findVerbs: (user: string) => Promise<Verb[]>;
  createOne: (fields: NewWordInput, user: string) => Promise<WordType | null>;
  updateOne: (
    update: UpdateWordInput,
    user: string
  ) => Promise<{ ok: boolean; value?: WordType | null }>;
  deleteOne: (id: string, user: string) => Promise<{ ok: boolean }>;
  updateStatistics: (
    data: UpdateStatisticsInput[],
    user: string
  ) => Promise<{ ok: boolean }>;
  selectWordsForGame: (
    filter: GameDataInput,
    config: GameSettings,
    user: string
  ) => Promise<WordType[]>;
  selectWordsForOptions: (
    gameType: GameType,
    language: Language,
    config: GameSettings,
    user: string
  ) => Promise<WordType[]>;
  getWordsForPractice: (
    filter: GameDataInput,
    config: GameSettings,
    user: string
  ) => Promise<WordType[]>;
}

export const WordModel: WordModelType = {
  async findById(id, user) {
    const word = await Word.findById(id).populate('tags').exec();
    if (user !== word?.user) {
      return null;
    }

    return word.toObject();
  },

  async findOne(filter, user) {
    const word = await Word.findOne({ ...filter, user })
      .populate('tags')
      .exec();
    return word?.toObject();
  },

  async findManyAndPaginate(filter, user) {
    const sortingFilter = getWordSortingFilters(
      filter?.sortBy || SortWordsBy.UpdatedAt,
      filter?.isReverseOrder
    );
    const filters = getWordsFilters(filter, user);
    const wordsCount = await Word.countDocuments(filters);

    const { limit, hasNext, skip, page } = calculatePaginationValues(
      filter?.limit,
      filter?.page,
      wordsCount
    );

    const words = await Word.find(filters)
      .sort(sortingFilter)
      .limit(limit)
      .skip(skip)
      .populate('tags')
      .exec();

    const result: PaginatedWords = {
      words,
      page,
      limit,
      hasNext,
      wordsCount
    };

    return result;
  },

  async createOne(fields, user) {
    const { uuid } = fields;
    let existing;
    if (uuid) {
      existing = await Word.findOne({
        uuid,
        user
      });
    }

    if (existing) {
      throw new UserInputError(`word with uuid ${uuid} is already added`);
    }
    const createdAt = Date.now();
    const updatedAt = createdAt;
    const stems = fields?.stems?.length ? fields.stems : [fields.name];
    const transcription = fields?.transcription || fields?.name;
    const isOffensive = !!fields?.isOffensive;
    const { tags } = fields;
    const tagsField = await filterTags({
      tags,
      language: fields.language,
      user
    });
    const spacedRepetition = DEFAULT_REPETITION_DATA;
    const word = await Word.create({
      ...fields,
      user,
      createdAt,
      updatedAt,
      stems,
      isOffensive,
      transcription,
      statistics: DEFAULT_STATISTICS,
      tags: tagsField,
      spacedRepetition
    });
    return word.toObject();
  },

  async updateOne(fields, user) {
    const updatedAt = Date.now();
    const word = await Word.findOne({ _id: fields.id, user });
    if (!word) {
      return { ok: false };
    }
    const language = word.language;
    const { tags } = fields;
    const tagsUpdate = await filterTags({ tags, language, user });
    const update = { ...fields, tags: tagsUpdate, updatedAt };
    const { acknowledged, modifiedCount } = await word.updateOne(update);
    const updatedWord = await Word.findById(word.id);
    const isOk = modifiedCount == 1 && acknowledged === true;
    return { ok: isOk, value: updatedWord?.toObject() };
  },

  async updateStatistics(data, user) {
    let isOk = true;
    data.forEach(async entry => {
      try {
        const word = await Word.findById(entry.id);
        if (!word || word.user !== user) {
          isOk = false;
          return;
        }
        if (!word?.statistics) {
          word.statistics = DEFAULT_STATISTICS;
        }

        const { gameType, tense, score, hasError = false } = entry;
        const currentStatistics = word?.statistics?.[gameType];

        if (!currentStatistics) {
          word.statistics[gameType] = DEFAULT_STATISTICS[gameType];
        }

        const newStatistics = {
          practicedTimes: 1,
          errorCount: currentStatistics?.errorCount || 0,
          successRate: currentStatistics?.successRate || 0,
          lastTimePracticed: Date.now()
        };

        if (hasError) {
          newStatistics.errorCount++;
          newStatistics.successRate = 0;
        }

        if (!hasError) {
          newStatistics.successRate++;
        }

        if (currentStatistics?.practicedTimes) {
          newStatistics.practicedTimes += currentStatistics.practicedTimes;
        }
        word.statistics[gameType] = newStatistics;

        if (!word?.spacedRepetition) {
          word.spacedRepetition = DEFAULT_REPETITION_DATA;
        }

        const spacedRepetition = word.spacedRepetition[gameType];

        const spacedRepetitionData =
          tense && isConjugationSpacedRepetitionData(spacedRepetition)
            ? spacedRepetition[tense]
            : (spacedRepetition as SpacedRepetitionData);

        const updatedSpacedepetition = updateSpacedRepetitionFields(
          spacedRepetitionData,
          score || Score.Hard
        );

        if (tense && gameType === Game.Conjugation) {
          word.spacedRepetition[gameType][tense] = updatedSpacedepetition;
        } else {
          // @ts-expect-error: figure out how to  define type dynamically here
          word.spacedRepetition[gameType] = updatedSpacedepetition;
        }
        await word.save();
      } catch (err) {
        console.error('saving statisticks failed', err);
        isOk = false;
      }
    });
    return { ok: isOk };
  },

  async deleteOne(id, user) {
    const result = { ok: false };
    const { deletedCount, acknowledged } = await Word.deleteOne({
      _id: id,
      user
    });
    if (acknowledged && deletedCount == 1) {
      result.ok = true;
    }
    return result;
  },

  async selectWordsForGame(filter, config, user) {
    const { tags, gameType, language, sortBy, isReverseOrder } = filter;

    const tagsFilters = getTagsFilter(tags);
    const gameFilters = getGameFilters(
      gameType,
      config.timesToLearn,
      sortBy === SortBy.MemoryRefresher,
      filter?.tense
    );

    const sortingFilters = getGameSortingFilter(
      gameType,
      sortBy,
      isReverseOrder
    );

    const words = await Word.find(
      { user, language, ...tagsFilters, ...gameFilters },
      PROJECTIONS[gameType]
    )
      .sort(sortingFilters)
      .limit(config.wordsPerGame)
      .exec();
    return words;
  },

  async selectWordsForOptions(gameType, language, config, user) {
    const optionProjections = OPTIONS_PROJECTIONS[gameType];
    const words = Word.find({ language, user }, optionProjections, {
      limit: config.optionsPerGame + DEFAULT_OPTIONS__SAMPLE_SIZE
    });
    return words;
  },

  async findVerbs(user) {
    const verbs = (await Word.find(
      {
        user,
        language: Language.Spanish,
        ...GAME_FILTERS[GameType.Conjugation]
      },
      'name id'
    )
      .sort({ name: 1 })
      .exec()) as Verb[];
    return verbs;
  },

  async getWordsForPractice(filter, config, user) {
    const { gameType, tense, language, tags } = filter;
    const now = new Date();
    let spacedRepetitionField = `spacedRepetition.${gameType}`;

    if (tense && gameType === Game.Conjugation) {
      spacedRepetitionField = `${spacedRepetitionField}.${tense}`;
    }
    const dateField = `${spacedRepetitionField}.nextReviewDate`;

    const tagsFilters = getTagsFilter(tags);
    const gameFilters = GAME_FILTERS?.[gameType] || {};
    const tenseFilter = getTenseFilter(gameType, tense);

    const words = await Word.find(
      {
        user,
        language,
        $or: [
          { [dateField]: { $lte: now } },
          { spacedRepetition: { $exists: false } }
        ],
        ...tagsFilters,
        ...gameFilters,
        ...tenseFilter
      },
      PROJECTIONS[gameType]
    )
      // priotitize new cards
      .sort({ [dateField]: -1 })
      .limit(config.wordsPerGame)
      .exec();
    return words;
  }
};
