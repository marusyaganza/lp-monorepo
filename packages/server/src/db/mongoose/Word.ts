import { Schema, model } from 'mongoose';
import { Word as WordCoreType } from 'generated/graphql';

const wordSchema = new Schema<WordCoreType>({
  uuid: { type: String },
  name: { type: String, required: true, immutable: true },
  defs: [{ def: { type: String, required: true }, examples: [String] }],
  user: { type: String, required: true },
  audioUrl: String,
  transcription: [String],
  imgUrl: String,
  isOffensive: Boolean,
  level: { type: String, enum: ['A1', 'A2', 'B1', 'B2', 'C1', 'C2'] },
  particle: String,
  stems: [String],
  tags: [String],
  additionalInfo: String
});

export const Word = model<WordCoreType>('Word', wordSchema);

