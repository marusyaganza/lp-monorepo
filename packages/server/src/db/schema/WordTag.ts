import { Schema, model } from 'mongoose';
import { WordTag as CoreWordTagType, Language } from '../../generated/graphql';

export interface WordTagType extends CoreWordTagType {
  user: string;
}

const wordTagSchema = new Schema<WordTagType>({
  text: { type: String, required: true },
  color: { type: String, required: true },
  user: { type: String, required: true },
  language: { type: String, enum: Language, required: true },
  desc: String
});

export const WordTag = model<WordTagType>('WordTag', wordTagSchema);
