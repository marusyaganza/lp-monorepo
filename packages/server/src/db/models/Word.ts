import { QueryOptions } from 'mongoose';
import { Word } from '../schema/Word';
import { WordTag } from '../schema/WordTag';
import { Game } from '../schema/Game';

import {
  Word as WordType,
  NewWordInput,
  UpdateStatisticsInput,
  Language,
  Game as GameType,
  SortBy,
  SortWordsBy,
  WordStatisticsField
} from '../../generated/graphql';
import { formatFilter, formatData } from '../helpers';

const STATISTICS_FIELD: WordStatisticsField = {
  lastTimePracticed: 0,
  practicedTimes: 0,
  errorCount: 0,
  successRate: 0
};

const DEFAULT_STATISTICS: Record<GameType, WordStatisticsField> = {
  [GameType.Audio]: STATISTICS_FIELD,
  [GameType.SelectDef]: STATISTICS_FIELD,
  [GameType.SelectWord]: STATISTICS_FIELD,
  [GameType.TypeWord]: STATISTICS_FIELD,
  [GameType.Conjugation]: STATISTICS_FIELD
};

type wordsFilter = {
  sortBy?: SortBy | SortWordsBy | 'updatedAt';
  language: Language;
  isReverseOrder: boolean;
  timesToLearn?: number | null;
  gameType?: GameType;
  user?: string;
  tags?: string[];
};

export interface WordModelType {
  findOne: (filter: Partial<WordType>) => Promise<WordType | null>;
  findMany: (
    filter: Partial<WordType>,
    projection: string,
    options?: QueryOptions
  ) => Promise<WordType[] | null>;
  findManyAndSort: (filter: wordsFilter) => Promise<WordType[] | null>;
  createOne: (fields: NewWordInput) => Promise<WordType | null>;
  updateOne: (
    fields: Partial<WordType> & Pick<WordType, 'id' | 'user'>
  ) => Promise<{ ok: boolean; value?: WordType | null }>;
  updateMany: (
    fields: Partial<WordType> & Pick<WordType, 'id'>[],
    user?: string
  ) => Promise<{ ok: boolean }>;
  deleteOne: (filter: {
    id: string;
    user?: string;
  }) => Promise<{ ok: boolean }>;
  updateStatistics: (
    data: UpdateStatisticsInput[],
    user?: string
  ) => Promise<{ ok: boolean }>;
}

export const WordModel: WordModelType = {
  async findOne(filter) {
    if (!filter?.user) {
      return null;
    }
    const word = await Word.findOne(formatFilter(filter))
      .populate('tags')
      .exec();
    return formatData(word);
  },

  async findMany(filter, projection, options) {
    if (!filter?.user) {
      return [];
    }
    const words = await Word.find(formatFilter(filter), projection, options)
      .populate('tags')
      .exec();
    return words;
  },

  // Check if criteria is defined, if so => sort by criteria, else use natural sorting
  async findManyAndSort(filter) {
    const {
      user,
      language = Language.English,
      sortBy,
      isReverseOrder,
      gameType,
      tags,
      timesToLearn = 5
    } = filter;

    if (!user) {
      return [];
    }

    let orderNum = isReverseOrder ? -1 : 1;

    // words with the highest number of errors should go first
    if (sortBy === SortBy.ErrorCount) {
      orderNum = -1 * orderNum;
    }
    let learningStatusFilter = {};
    let sort: Record<string, number> = { $natural: orderNum };

    let gameFilter;

    if (gameType === GameType.Conjugation) {
      gameFilter = { conjugation: { $ne: null } };
    }

    if (gameType === GameType.Audio) {
      gameFilter = { audioUrl: { $ne: '' } };
    }

    let tagsFilters: { tags: string }[] = [];
    if (tags?.length) {
      tagsFilters = tags?.map(tag => {
        return { tags: tag };
      });
    }
    // TODO: refactor this part
    // Select words that are not learned or have been practiced without error less than 5 times in a row
    if (gameType && sortBy !== SortBy.MemoryRefresher) {
      learningStatusFilter = {
        $and: [
          { isLearned: { $ne: true } },
          ...tagsFilters,
          gameFilter,
          {
            $or: [
              { [`statistics.${gameType}.successRate`]: { $lt: timesToLearn } },
              { [`statistics.${gameType}.successRate`]: null }
            ]
          }
        ]
      };
    } else {
      if (tagsFilters?.length) {
        learningStatusFilter = {
          $and: tagsFilters
        };
      }
    }

    if (sortBy) {
      let propName: string = sortBy;
      if (gameType) {
        propName = `statistics.${gameType}.${sortBy}`;
      }

      if (sortBy === SortBy.MemoryRefresher) {
        learningStatusFilter = {
          $or: [
            { [`statistics.${gameType}.successRate`]: { $gte: timesToLearn } },
            { isLearned: true }
          ]
        };
      } else {
        sort = { [propName]: orderNum };
      }
    }

    const filters = {
      user,
      language,
      ...learningStatusFilter
    };

    const words = await Word.find(filters)
      // @ts-ignore
      .sort(sort)
      .populate('tags')
      .exec();

    return words;
  },

  async createOne(fields) {
    const createdAt = Date.now();
    const updatedAt = createdAt;
    const stems = fields?.stems?.length ? fields.stems : [fields.name];
    const transcription = fields?.transcription || fields?.name;
    const isOffensive = !!fields?.isOffensive;
    const word = await Word.create({
      ...fields,
      createdAt,
      updatedAt,
      stems,
      isOffensive,
      transcription,
      statistics: DEFAULT_STATISTICS
    });
    return formatData(word);
  },

  async updateOne(fields) {
    const updatedAt = Date.now();
    const word = await Word.findOne({ _id: fields.id, user: fields.user });
    if (!word) {
      return { ok: false };
    }
    const language = word.language;
    const { user, tags } = fields;
    let existingTags = await WordTag.find({ user, language }, 'id');
    existingTags = existingTags.map(tag => tag.id);
    // @ts-ignore
    const tagsUpdate = tags?.filter(tag => existingTags.includes(tag));
    const update = { ...fields, tags: tagsUpdate, updatedAt };
    const { acknowledged, modifiedCount } = await word.updateOne(update);
    const updatedWord = await Word.findById(word.id);
    const isOk = modifiedCount == 1 && acknowledged === true;
    return { ok: isOk, value: formatData(updatedWord) };
  },

  // This method do not have a practical use for now
  async updateMany(data, user) {
    let isOk = true;
    const updatedAt = Date.now();
    data.forEach(async entry => {
      const update = { ...entry, updatedAt };
      const result = await Word.findOneAndUpdate(
        { _id: entry.id, user },
        update,
        { includeResultMetadata: true, new: true }
      );
      isOk = isOk && result?.ok == 1 && result?.value !== null;
    });

    return { ok: isOk };
  },

  async updateStatistics(data) {
    let isOk = true;
    data.forEach(async entry => {
      try {
        const word = await Word.findById(entry.id);
        if (!word) {
          isOk = false;
          return;
        }
        if (!word?.statistics) {
          word.statistics = DEFAULT_STATISTICS;
        }

        const { gameType, hasError = false, isLearned = false } = entry;
        const gameConfig = await Game.findOne({ type: gameType });
        const timesToLearn = gameConfig?.timesToLearn || 5;

        const currentStatistics = word?.statistics?.[gameType];

        if (!currentStatistics) {
          word.statistics[gameType] = DEFAULT_STATISTICS[gameType];
        }

        const newStatistics = {
          practicedTimes: 1,
          errorCount: currentStatistics?.errorCount ?? 0,
          successRate: currentStatistics?.successRate ?? 0,
          lastTimePracticed: Date.now()
        };

        if (hasError) {
          newStatistics.errorCount++;
          newStatistics.successRate = 0;
        }

        if (!hasError) {
          newStatistics.successRate++;
        }

        if (isLearned) {
          newStatistics.successRate = timesToLearn;
        }

        if (currentStatistics?.practicedTimes) {
          newStatistics.practicedTimes += currentStatistics.practicedTimes;
        }

        word.statistics[gameType] = newStatistics;
        await word.save();
      } catch (err) {
        console.error('saving statisticks failed', err);
        isOk = false;
      }
    });
    return { ok: isOk };
  },

  async deleteOne(filter) {
    const result = { ok: false };
    const { deletedCount, acknowledged } = await Word.deleteOne({
      _id: filter.id,
      user: filter.user
    });
    if (acknowledged && deletedCount == 1) {
      result.ok = true;
    }
    return result;
  }
};
