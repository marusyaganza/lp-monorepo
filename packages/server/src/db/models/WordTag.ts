import { formatData, formatFilter } from '../helpers';
import { Word } from '../schema/Word';
import { WordTag, WordTagType } from '../schema/WordTag';
import {
  WordTagInput,
  UpdateWordTagInput,
  WordTag as TagType
} from 'generated/graphql';

export interface WordTagModelType {
  findOne: (filter: Partial<WordTagType>) => Promise<WordTagType | null>;
  findMany: (filter: Partial<WordTagType>) => Promise<WordTagType[] | null>;
  createOne: (fields: WordTagInput) => Promise<WordTagType | null>;
  deleteOne: (filter: {
    id: string;
    user?: string;
  }) => Promise<{ ok: boolean; value?: TagType }>;
  updateOne: (
    fields: Partial<UpdateWordTagInput> & Pick<WordTagType, 'id' | 'user'>
  ) => Promise<{ ok: boolean; value: WordTagType | null }>;
}

export const WordTagModel: WordTagModelType = {
  async createOne(fields) {
    const tag = await WordTag.create(fields);
    return formatData(tag);
  },

  async findOne(filter) {
    if (!filter?.user) {
      return null;
    }
    const tag = await WordTag.findOne(formatFilter(filter));
    return formatData(tag);
  },

  async findMany(filter) {
    if (!filter?.user) {
      return [];
    }
    const tags = await WordTag.find(formatFilter(filter));
    return tags;
  },
  async updateOne(fields) {
    const { ok, value } = await WordTag.findOneAndUpdate(
      { _id: fields.id, user: fields.user },
      fields,
      { includeResultMetadata: true, new: true }
    );
    const isOk = ok == 1 && value !== null;
    return { ok: isOk, value: formatData(value) };
  },

  async deleteOne(filter) {
    const { id, user } = filter;

    let result = { ok: false };

    const tag = await WordTag.findOne({
      _id: filter.id,
      user: filter.user
    });

    if (!tag) {
      return result;
    }

    const wordsWithTag = await Word.find(
      {
        tags: id,
        user: user
      },
      'tags'
    );

    const tagId = tag._id;

    wordsWithTag.forEach(async wordWithTag => {
      if (wordWithTag?.tags?.length) {
        // @ts-ignore
        wordWithTag.tags = wordWithTag.tags.filter(tag => tag != tagId);
        await wordWithTag.save();
      }
    });

    await tag.deleteOne();
    // @ts-ignore
    result = { ok: true, value: tag };

    return result;
  }
};
