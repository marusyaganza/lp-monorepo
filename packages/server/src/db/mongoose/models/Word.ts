import { Word } from '../Word';
import { Word as WordType } from '../../../generated/graphql';
import { ModelType } from '../../';
import { formatFilter, formatData } from '../helpers';
// import { UserInputError } from '../../../utils/apolloCustomErrors';

export const WordModel: ModelType<WordType> = {
  // @ts-ignore
  async findOne(filter = {}) {
    if (!filter?.user) {
      return null;
    }
    const word = await Word.findOne(formatFilter(filter));
    return formatData(word);
  },

  // @ts-ignore
  async findMany(filter = {}) {
    if (!filter?.user) {
      return [];
    }
    const words = await Word.find(formatFilter(filter));
    return words;
  },

  // @ts-ignore
  async createOne(fields) {
    const createdAt = Date.now();
    const word = await Word.create({ ...fields, createdAt });
    return formatData(word);
  },

  // @ts-ignore
  async updateOne(fields) {
    const update = { ...fields };
    delete update.id;
    delete update.user;
    const { ok, value } = await Word.findOneAndUpdate(
      { _id: fields.id, user: fields.user },
      update,
      { rawResult: true, new: true }
    );
    const isOk = ok == 1 && value !== null;
    return { ok: isOk, value: formatData(value) };
  },

  async deleteOne(filter) {
    const result = { ok: false };
    const { deletedCount, acknowledged } = await Word.deleteOne({
      _id: filter.id,
      user: filter.user
    });
    if (acknowledged && deletedCount == 1) {
      result.ok = true;
    }
    return result;
  }
};
