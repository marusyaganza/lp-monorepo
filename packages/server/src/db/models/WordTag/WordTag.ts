import { Word } from '../../schema/Word';
import { WordTag } from '../../schema/WordTag';
import {
  WordTagInput,
  UpdateWordTagInput,
  WordTag as TagType
} from '../../../generated/graphql';

type TagDeleteResult = {
  ok: boolean;
  value: TagType | null;
};

export interface WordTagModelType {
  findOne: (filter: Partial<TagType>, user: string) => Promise<TagType | null>;
  findMany: (filter: Partial<TagType>, user: string) => Promise<TagType[]>;
  createOne: (fields: WordTagInput, user: string) => Promise<TagType | null>;
  deleteOne: (id: string, user: string) => Promise<TagDeleteResult>;
  updateOne: (
    fields: Partial<UpdateWordTagInput> & Pick<TagType, 'id'>,
    user: string
  ) => Promise<{ ok: boolean; value: TagType | null | undefined }>;
}

export const WordTagModel: WordTagModelType = {
  async createOne(fields, user) {
    const tag = await WordTag.create({ ...fields, user });
    return tag.toObject();
  },

  async findOne(filter, user) {
    let tag;
    if ('id' in filter) {
      tag = await WordTag.findById(filter.id);
    } else {
      tag = await WordTag.findOne(filter);
    }
    if (tag?.user !== user) {
      return null;
    }
    return tag.toObject();
  },

  async findMany(filter, user) {
    const tags = await WordTag.find({ ...filter, user });
    return tags || [];
  },

  async updateOne(fields, user) {
    const result = await WordTag.findOneAndUpdate(
      { _id: fields.id, user },
      fields,
      { includeResultMetadata: true, new: true }
    );
    const { ok, value } = result;
    const isOk = ok == 1 && value !== null;
    return { ok: isOk, value: value?.toObject() };
  },

  async deleteOne(id, user) {
    let result: TagDeleteResult = { ok: false, value: null };

    const tag = await WordTag.findOne({
      _id: id,
      user
    });

    if (!tag) {
      return result;
    }

    const wordsWithTag = await Word.find(
      {
        tags: id,
        user: user
      },
      'tags'
    );

    const tagId = tag.id.toString();

    wordsWithTag.forEach(async wordWithTag => {
      if (wordWithTag?.tags?.length) {
        wordWithTag.tags = wordWithTag.tags.filter(tag => tag != tagId);
        await wordWithTag.save();
      }
    });

    await tag.deleteOne();
    result = { ok: true, value: tag.toObject() };

    return result;
  }
};
