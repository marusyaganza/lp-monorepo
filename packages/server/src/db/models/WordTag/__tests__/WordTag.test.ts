import { WordTagModel } from '../WordTag';
import { WordModel } from '../../Word/Word';

import {
  connectToDb,
  disconnectFromDb,
  dropDb,
  baseSnapshotConfig
} from '../../../../tests/helpers';
import { testData } from '../../../../tests/mocks/dbTestData';

const mockUser = { id: '6480560e8cad1841ed6b4011', role: 'MEMBER' };
const mockId = '6480653341e6f90377d19cfb';

describe('WordTagModel', () => {
  beforeEach(async () => {
    await connectToDb();
    await dropDb();
  });

  afterEach(async () => {
    await dropDb();
    await disconnectFromDb();
  });

  test('createOne', async () => {
    const result = await WordTagModel.createOne(
      testData.createTagInput,
      mockUser.id
    );
    expect(result).toMatchSnapshot(baseSnapshotConfig);
  });

  test('findOne', async () => {
    const tag = await WordTagModel.createOne(
      testData.createTagInput,
      mockUser.id
    );
    const id = tag?.id as string;
    await WordTagModel.createOne(testData.createTagInput, mockId);
    const result = await WordTagModel.findOne(
      {
        id
      },
      mockUser.id
    );
    const result2 = await WordTagModel.findOne(
      {
        text: testData.createTagInput.text
      },
      mockUser.id
    );
    expect(result2).not.toBeNull();
    // @ts-expect-error: testing possible empty args
    expect(result.id).toEqual(result2.id);
    // @ts-expect-error: testing possible empty args
    expect(result.user).toEqual(result2.user);
    expect(result).toMatchSnapshot(baseSnapshotConfig);
  });

  test('findOne should return null if user id is not provided', async () => {
    const tag = await WordTagModel.createOne(
      testData.createTagInput,
      mockUser.id
    );
    const id = tag?.id as string;

    // @ts-expect-error: testing empty args case
    const result = await WordTagModel.findOne({
      id
    });
    // @ts-expect-error: testing empty args case
    const result2 = await WordTagModel.findOne({
      uuid: testData.createTagInput.uuid
    });
    expect(result).toBeNull();
    expect(result2).toBeNull();
  });

  test('findMany', async () => {
    await WordTagModel.createOne(testData.createTagInput, mockUser.id);
    await WordTagModel.createOne(testData.createTagInput2, mockUser.id);
    await WordTagModel.createOne(testData.createTagInput, mockId);

    // @ts-expect-error: testing empty args case
    const resultForUnknowUser = await WordTagModel.findMany({});
    const result = await WordTagModel.findMany({}, mockUser.id);
    expect(result).toHaveLength(2);
    expect(resultForUnknowUser).toHaveLength(0);
  });

  test('updateOne', async () => {
    const tag = await WordTagModel.createOne(
      testData.createTagInput,
      mockUser.id
    );
    const id = tag?.id as string;

    const tag2 = await WordTagModel.createOne(
      testData.createTagInput2,
      mockUser.id
    );

    const result = await WordTagModel.updateOne(
      {
        ...testData.updateTagInput,
        id
      },
      mockUser.id
    );

    expect(result.ok).toBe(true);
    expect(result.value).toMatchSnapshot(baseSnapshotConfig);

    const searchTResult = await WordTagModel.findOne(
      {
        id
      },
      mockUser.id
    );

    expect(result.value).toEqual(searchTResult);

    // @ts-expect-error: testing empty args case
    const resultForUnknowUser = await WordTagModel.updateOne({
      ...testData.updateTagInput,
      id: tag2?.id
    });
    expect(resultForUnknowUser).toEqual({ ok: false, value: undefined });
  });

  test('deleteOne', async () => {
    const tag = await WordTagModel.createOne(
      testData.createTagInput,
      mockUser.id
    );
    const id = tag?.id as string;
    await WordTagModel.createOne(testData.createTagInput2, mockUser.id);
    await WordTagModel.createOne(testData.createTagInput, mockId);

    const wordWithTag = await WordModel.createOne(
      {
        ...testData.createWordInput,
        tags: [id]
      },
      mockUser.id
    );

    expect(wordWithTag?.tags?.length).toEqual(1);
    const wordWithTagId = wordWithTag?.id as string;

    const result = await WordTagModel.deleteOne(id, mockUser.id);

    const wordAfter = await WordModel.findById(wordWithTagId, mockUser.id);

    expect(result).toMatchObject({ ok: true, value: tag });
    expect(wordAfter?.tags).toEqual([]);
    const searchResult = await WordTagModel.findOne(
      {
        id
      },
      mockUser.id
    );
    const tags = await WordTagModel.findMany({}, mockUser.id);
    expect(tags).toHaveLength(1);
    expect(searchResult).toBeNull();

    const existingTags = await WordTagModel.findMany({}, mockUser.id);

    // @ts-expect-error: testing empty args case
    const resultForUnknowUser = await WordTagModel.deleteOne(tag.id);
    expect(resultForUnknowUser).toEqual({ ok: false, value: null });

    // @ts-expect-error: testing empty args case
    const resultForUnexistingTag = await WordTagModel.deleteOne(mockId);

    expect(resultForUnexistingTag).toEqual({ ok: false, value: null });
    expect(existingTags).toHaveLength(1);
  });
});
