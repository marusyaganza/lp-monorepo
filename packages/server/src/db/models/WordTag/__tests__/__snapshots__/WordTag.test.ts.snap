// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WordTagModel createOne 1`] = `
{
  "_id": Any<Object>,
  "color": "#f0f0f0",
  "desc": "Tag description",
  "id": Any<String>,
  "language": "ENGLISH",
  "text": "Tag text",
  "user": "6480560e8cad1841ed6b4011",
}
`;

exports[`WordTagModel findOne 1`] = `
{
  "_id": Any<Object>,
  "color": "#f0f0f0",
  "desc": "Tag description",
  "id": Any<String>,
  "language": "ENGLISH",
  "text": "Tag text",
  "user": "6480560e8cad1841ed6b4011",
}
`;

exports[`WordTagModel updateOne 1`] = `
{
  "_id": Any<Object>,
  "color": "#f0f0f1",
  "desc": "Tag description was updated",
  "id": Any<String>,
  "language": "ENGLISH",
  "text": "Updated Tag text",
  "user": "6480560e8cad1841ed6b4011",
}
`;
